src/EndpointManager.sol

```diff
@@ -241,34 +313,62 @@ abstract contract EndpointManager is
-        _consumeOutboundAmount(amount);
-
-        if (_mode == Mode.LOCKING) {
-            // use transferFrom to pull tokens from the user and lock them
-            // query own token balance before transfer
-            uint256 balanceBefore = getTokenBalanceOf(_token, address(this));
-
-            // transfer tokens
-            IERC20(_token).safeTransferFrom(msg.sender, address(this), amount);
-
-            // query own token balance after transfer
-            uint256 balanceAfter = getTokenBalanceOf(_token, address(this));
-
-            // correct amount for potential transfer fees
-            amount = balanceAfter - balanceBefore;
-        } else {
-            // query sender's token balance before transfer
-            uint256 balanceBefore = getTokenBalanceOf(_token, msg.sender);
-
-            // call the token's burn function to burn the sender's token
-            ERC20Burnable(_token).burnFrom(msg.sender, amount);
-
-            // query sender's token balance after transfer
-            uint256 balanceAfter = getTokenBalanceOf(_token, msg.sender);
+        if (shouldCheckRateLimit) {
+            // Lock/burn tokens before checking rate limits
+            if (_mode == Mode.LOCKING) {
+                // use transferFrom to pull tokens from the user and lock them
+                // query own token balance before transfer
+                uint256 balanceBefore = getTokenBalanceOf(_token, address(this));
+
+                // transfer tokens
+                IERC20(_token).safeTransferFrom(msg.sender, address(this), amount);
+
+                // query own token balance after transfer
+                uint256 balanceAfter = getTokenBalanceOf(_token, address(this));
+
+                // correct amount for potential transfer fees
+                amount = balanceAfter - balanceBefore;
+            } else {
+                // query sender's token balance before transfer
+                uint256 balanceBefore = getTokenBalanceOf(_token, msg.sender);
+
+                // call the token's burn function to burn the sender's token
+                ERC20Burnable(_token).burnFrom(msg.sender, amount);
+
+                // query sender's token balance after transfer
+                uint256 balanceAfter = getTokenBalanceOf(_token, msg.sender);
+
+                // correct amount for potential burn fees
+                amount = balanceAfter - balanceBefore;
+            }
+
+            // now check rate limits
+            bool didConsumeAmount = _consumeOutboundAmount(amount);
+            if (shouldQueue && !didConsumeAmount) {
+                // queue up and return
+                uint64 queueSequence = useOutboundQueueSequence();
+
+                _getOutboundQueueStorage()[queueSequence] = OutboundQueuedTransfer({
+                    amount: amount,
+                    recipientChain: recipientChain,
+                    recipient: recipient,
+                    txTimestamp: block.timestamp
```

Comment:
would it better to use ` _getOutboundLimitParamsStorage().lastTxTimestamp` here? or add an `assert` to check that block.timestamp is equivalent to  ` _getOutboundLimitParamsStorage().lastTxTimestamp` when the transfer is added to the queue? if `didConsume` is true, this variable is set here: https://github.com/wormhole-foundation/example-native-token-transfers/tree/0099b82e1daf519efdf1b1748aa014e3cf7b8356/src/EndpointManager.sol

Reply:
> But it's an interesting callout. Should we update only  _getOutboundLimitParamsStorage().lastTxTimestamp to be block.timestamp when we queue up a transfer? (That would effectively reset the rate-limit recharge)

That would basically make it a stepwise linear recharge right? Where the gradient decreases each time you queue up a transfer? Imo I don't think we should do that unless there's a strong argument for it

Reply:
seems like it's stepwise constant for each set of transfers added to the queue until the capacity is reached. agreed that this is probably not a good idea